src/node/automated-browser-testing.js

```diff
@@ -0,0 +1,260 @@
+        }
+      }
+    ];
+
+    const discoverableBrowsers = [];
+    browserExecutables.forEach(browserInfo => {
+      try {
+        const executablePath = browserInfo.getExecutableName();
+        const seleniumOptions = browserInfo.seleniumOptions;
+
+        if (seleniumOptions.setChromeBinaryPath) {
+          seleniumOptions.setChromeBinaryPath(executablePath);
+        } else if (seleniumOptions.setBinary) {
+          seleniumOptions.setBinary(executablePath);
+        } else {
+          throw new Error('Unknown selenium options object');
+        }
+
+        browserInfo.getSeleniumDriver = () => {
+          return new webdriver
+            .Builder()
+            .forBrowser(browserInfo.seleniumBrowserId)
+            .setChromeOptions(browserInfo.seleniumOptions)
+            .setFirefoxOptions(browserInfo.seleniumOptions)
+            .build();
+        };
+
+        discoverableBrowsers.push(browserInfo);
+      } catch (err) {
+        // Unable to find the browser - chances are this isn't an error
+        // but instead the browser is just not installed
+      }
+    });
+
+    return discoverableBrowsers;
+  }
+
+  runMochaTests(browserName, driver, url) {
+    // The driver methods are wrapped in a new promise because the
+    // selenium-webdriver API seems to using some custom promise
+    // implementation that has slight behaviour differences.
+    return new Promise((resolve, reject) => {
+      driver.get(url)
+      .then(() => {
+        return driver.executeScript('return window.navigator.userAgent;');
+      })
+      .then(userAgent => {
+        // This is just to help with debugging so we can get the browser version
+        console.log('        [' + browserName + ' UA]: ' + userAgent);
+      })
+      .then(() => {
+        // We get webdriver to wait until window.testsuite.testResults is defined.
+        // This is set in the in browser mocha tests when the tests have finished
+        // successfully
+        return driver.wait(function() {
+          return driver.executeScript('return ((typeof window.testsuite !== ' +
+            '\'undefined\') && window.testsuite.testResults !== ' +
+            '\'undefined\');');
+        });
+      })
+      .then(() => {
+        // This simply retrieves the test results from the inbrowser mocha tests
+        return driver.executeScript('return window.testsuite.testResults;');
+      })
+      .then(testResults => {
+        // Resolve the outer promise to get out of the webdriver promise chain
+        resolve(testResults);
+      })
+      .thenCatch(reject);
+    });
+  }
+
+  killWebDriver(driver) {
+    return new Promise(resolve => {
+      // Suggested as fix to 'chrome not reachable'
+      // http://stackoverflow.com/questions/23014220/webdriver-randomly-produces-chrome-not-reachable-on-linux-tests
+      const timeoutGapCb = function() {
+        setTimeout(resolve, 2000);
+      };
+
+      if (driver === null) {
+        return timeoutGapCb();
```

Comment:
Why would you need to wait 2 seconds before resolving when `driver` is `null`? ("Selenium WebDriver is weird." is an acceptable answer.)